src/nvim/screen.c

```diff
@@ -3338,6 +3346,16 @@ win_line (
             char_attr = hl_combine_attr(spell_attr, char_attr);
         }
 
+        if (has_bufhl && v > 0) {
+          bufhl_attr = bufhl_get_attr(&bufhl_info, (colnr_T)v);
+          if( bufhl_attr != 0) {
+            if (!attr_pri)
+              char_attr = hl_combine_attr(char_attr, bufhl_attr);
+            else
+              char_attr = hl_combine_attr(bufhl_attr, char_attr);
+          }
+        }
```

Comment:
Was some time ago, I thought about simplifying the combination logic but didn't dare given the exponential cases of combination to test. I did it this way because it would work "as least as well" as spell attrs. Your suggestion would mean, IMHO, leaking parts of this complex logic to a separate  function, and could rather make it harder to reason what happens, say if an upstream patch includes a new highlighting source. ( I mean it might give the impression of "separating concerns" but that is not actually possible without deeper analysis/refactor of the highlight combining logic)